index.js

```diff
@@ -1,130 +1,119 @@
+    this.logger = options.logger || function () {}
   }
 
-  const opts = {
-    json: true,
-    method: 'POST',
-    uri: this.addr,
-    timeout: options.timeout || this.timeout,
-    body: body
+  prepareRequest (method, params, options) {
+    const id = options.async ? null : uid(16)
+    return {
+      params: Array.isArray(params) ? params : [ params ],
+      jsonrpc: '2.0',
+      method,
+      id
+    }
   }
 
-  const startTime = new Date()
-  return new Promise(function (resolve, reject) {
-    self.request(opts, function (err, res) {
-      const endTime = new Date()
-      const duration = endTime - startTime
-      self.log(method, params, duration, res, err)
+  makeHTTPRequest (body, options, fn) {
+    const requestOptions = {
+      json: true,
+      method: 'POST',
+      timeout: options.timeout || this.timeout,
+      uri: url.format(this.address),
+      body
+    }
+
+    request.post(requestOptions, function (err, res, body) {
+      body = body || {}
 
       if (err) {
-        reject(err)
-      } else {
-        resolve(res)
+        debug('error for %s: %s', options.id, err.message)
+        return fn(err)
       }
-    })
-  })
-}
-
-/**
- * Make the request.
- *
- * @param {Object} opts
- * @param {Function} fn
- * @api private
- */
-
-Client.prototype.request = function (opts, fn) {
-  const id = opts.id
-  debug('request %j', opts.body)
-  request.post(opts, function (err, res, body) {
-    body = body || {}
-
-    if (err) {
-      debug('error for %s: %s', id, err.message)
-      return fn(err)
-    }
 
-    if (body.error) {
-      if (typeof body.error === 'object') {
-        const e = new Error(body.error.message)
-        e.code = body.error.code
-        e.data = body.error.data
-        debug('error for %s: %s', id, e.message)
-        return fn(e)
+      if (body.error) {
+        if (typeof body.error === 'object') {
+          const e = new Error(body.error.message)
+          e.code = body.error.code
+          e.data = body.error.data
+          debug('error for %s: %s', options.id, e.message)
+          return fn(e)
+        }
+
+        // XXX: why do we do this?
+        if (body.error !== 'not found') {
+          debug('error for %s: %s', options.id, body.error)
+          return fn(new Error(body.error))
+        }
       }
 
-      // XXX: why do we do this?
-      if (body.error !== 'not found') {
-        debug('error for %s: %s', id, body.error)
-        return fn(new Error(body.error))
-      }
-    }
+      debug('success %s: %j', options.id, body.result || {})
+      fn(null, body.result)
+    })
+  }
 
-    debug('success %s: %j', id, body.result || {})
-    fn(null, body.result)
-  })
-}
+  makeTCPRequest (body, options, fn) {
+    fn = once(fn)
+    let response = null
+    const socket = net.connect(this.address.port, this.address.hostname)
+    socket.setTimeout(options.timeout || this.timeout)
+    socket.on('error', fn)
+    socket.pipe(split(JSON.parse))
```

Comment:
@stephenmathieson Judging by the `makeTCPRequest` function, call wouldn't receive the result if the socket didn't close. Or, in case a timeout is reached, that way socket would close too.

How does jsonrpc server handle TCP connections? Does it close the connection after replying?